### DISCUSSION TITLE
Prettierの設定の手順調査&設定を考える

### DISCUSSION THREAD

**Discussion by pss-aileen** [first-contributions-ja/first-contributions-ja.github.io#69] [90: 知識・知見の共有]
まともに設定したことがないので調査する

**Comment by pss-aileen** [comment]
## したいこと
- 保存するたびに自動成形してほしい
- フォーマットのルールを決める

**Comment by pss-aileen** [comment]
## そもそも Prettier とは
- [プリティアーとは何ですか? · より美しく](https://prettier.io/docs/en/)
- コードフォーマッター
- サポートしている言語？
  - JavaScript、Vue、TypeScript、CSS、HTML、JSON、Markdown、YAML

**Reply by pss-aileen** [reply]
## Prettier と Linter の違い
- たぶん、PrettierとESLint的なやつの違い
- Prettier: 書式設定
- Linter: バグをキャッチ

**Comment by pss-aileen** [comment]
## 一般的な？設定項目
- printWidth
- tabWidth
- semi
- singleQuote
- 

### 参考
- [Options · Prettier](https://prettier.io/docs/en/options.html#print-width)
- [【VSCode】Prettierの使い方＆おすすめ設定を紹介｜北海道札幌市のホームページ・Web制作・システム開発会社｜株式会社マーベリックス](https://ma-vericks.com/blog/vscode-prettier/)
- [【自動整形】VSCodeでPrettierを使う方法【設定必要です】 | RalaCode](https://ralacode.com/blog/post/vscode-prettier/)

**Reply by pss-aileen** [reply]
`Format On Save` にしたらたぶん保存時にやってくれる

**Reply by pss-aileen** [reply]
`Default Formatter` を Prettier にする設定がいるらしい

**Reply by pss-aileen** [reply]
自分のローカルでチェックしたわけではないがい、デフォルトの設定は以下のような感じらしい

```js
{
“printWidth”: 80,
“tabWidth”: 2,
“useTabs”: false,
“semi”: true,
“singleQuote”: false,
“quoteProps”: “as-needed”,
“jsxSingleQuote”: false,
“trailingComma”: “none”,
“bracketSpacing”: true,
“jsxBracketSameLine”: false,
“arrowParens”: “avoid”,
“rangeStart”: 0,
“rangeEnd”: Infinity,
“parser”: “none”,
“filepath”: “none”,
“requirePragma”: false,
“insertPragma”: false,
“proseWrap”: “preserve”,
“htmlWhitespaceSensitivity”: “css”,
“vueIndentScriptAndStyle”: false,
“endOfLine”: “auto”,
}
```

[Prettier – Code formatter 【インストール・設定・使い方・機能しない時の対処方】](https://digitor.jp/textbook/vscode-prettier/)

**Comment by pss-aileen** [comment]
## インストールしてみる

**Reply by pss-aileen** [reply]
### インストール

```bash
npm install --save-dev --save-exact prettier
```
インストールしたらなぜかDependenciesの順番変わって謎...
もしかして早速適応された！？わかんない

![スクリーンショット 2024-03-20 14 27 28](https://github.com/first-contributions-ja/first-contributions-ja.github.io/assets/140990100/7a029040-bf2f-41cd-b6ca-0d0178bd7389)

**Reply by pss-aileen** [reply]
### prettierrcファイルを作って、エディタやツールにPrettierを使うことを知らせる
```bash
node --eval "fs.writeFileSync('.prettierrc','{}\n')"
```

![スクリーンショット 2024-03-20 14 30 19](https://github.com/first-contributions-ja/first-contributions-ja.github.io/assets/140990100/f202a195-a55c-4242-b4b8-ff77355afa14)

生成された、中身はこんな感じ

**Reply by pss-aileen** [reply]
### .prettierignore でフォーマットしないファイル・フォルダを指定する

- .prettierignore を root に作成

とりあえず以下のように記述

```markdown
# Ignore artifacts:
.next
docs
```
artifacts: 人工物

**Reply by pss-aileen** [reply]
でも！
- .gitignoreで指定されたところもフォーマットしない
- .eslintignore に基づいてignoreすることもできるっぽい

**Reply by pss-aileen** [reply]
とりあえずフォーマットしてみる
```
npx prettier . --write
```

**Reply by pss-aileen** [reply]
変わってほしくないところまで変わったので、変わらなくて良いものは変更を取り消して
![スクリーンショット 2024-03-20 14 42 00](https://github.com/first-contributions-ja/first-contributions-ja.github.io/assets/140990100/f651aa3c-cdc3-4b5c-b0f9-b252ffcefff7)

.prettierignore 追記

```markdown
# Ignore artifacts:
.next
docs
.github
```

**Reply by pss-aileen** [reply]
フォーマットの結果、`;` が追加されたり、改行が調整された

**Reply by pss-aileen** [reply]
> CI セットアップがある場合は、その一部として次のコマンドを実行して、全員が Prettier を実行していることを確認します。これにより、マージの競合やその他のコラボレーションの問題が回避されます。

```bash
npx prettier . --check
```

これでフォーマットができているか確認できるらしい
ただちょっとよくわからない

**Reply by pss-aileen** [reply]
以上でインストールは終了らしい。
ESLintとの連携などもできるらしいが一旦ここではおいておく

**Reply by pss-aileen** [reply]
package.jsonのscriptsに
```json
    "format": "prettier . --write"
```
を追加

一応動くか確認、OKだった

**Comment by pss-aileen** [comment]
## オプションについて調べる

**Reply by pss-aileen** [reply]
資料集め
- [ESLint + Prettierを導入したTypeScript開発環境](https://zenn.dev/big_tanukiudon/articles/c1ab3dba7ba111)
-

**Comment by pss-aileen** [comment]
## .prettierrcを変更してみる

**Reply by pss-aileen** [reply]
```json
{
  "semi": false,
  "singleQuote": true,
  "tabWidth": 2
}
```
セミコロン消えた、シングルクオーとになった、タブが2になったらしい（変化を感じない、たぶんもとが2だからかな）

**Reply by pss-aileen** [reply]
なので↑は設定しないことにしてみた

**Reply by pss-aileen** [reply]
うん、デフォルトのタブ2だった
https://prettier.io/docs/en/options#tab-width

**Comment by pss-aileen** [comment]
## 保存時にフォーマットして欲しい

[VSCodeでプロジェクトの設定ファイルを準備する](https://zenn.dev/youfuku/articles/7f1c8aaf1cc320)

**Reply by pss-aileen** [reply]
- .vscode ディレクトリ作成
- settings.json を作成

**Reply by pss-aileen** [reply]
```
{
  "editor.defaultFormatter": "esbenp.prettier-vscode", // フォーマッターをPrettierにする
  "editor.formatOnSave": true // ファイル保存時にフォーマットを実行
}
```

**Reply by pss-aileen** [reply]
ファイルを保存する時にフォーマットが走るようになった

**Comment by kazzyfrog** [comment]
関連することとして、以下貼っておきます：

- [prettier-plugin-tailwindcss (Prettierプラグイン)](https://techracho.bpsinc.jp/hachi8833/2023_11_08/136255#bp3)を導入すれば、推奨されるtailwindクラスの順序に従ってそれらを自動的に並べ替えることが可能！
- Github ActionsでPull Request作成時などのタイミングで、ESLint, Prettierなどを強制実行してコード品質のを保つことも可能！
    - https://blog.uzumaki-inc.jp/eslint-prettiervscode-github-actions
    - https://zenn.dev/ampersand/articles/e45c9a2c0af831

**Reply by pss-aileen** [reply]
ありがとうございます！！
